text/0000-declarative-shadow-dom-polyfill.md

```diff
@@ -0,0 +1,346 @@
+status: DRAFTED
+created_at: 2022-08-25
+updated_at: 2022-08-25
+pr: 
+---
+
+# Declarative ShadowDOM polyfill
+
+## Summary
+
+This RFC describes and compares different methods to polyfill [declarative ShadowDOM spec](https://web.dev/declarative-shadow-dom/). It discusses how to integrate them into LWC-SSR. Finally, it also can be used as a guide for application containers to decide which polyfill fits best their use case.
+
+## Basic example
+
+Let the following snippet represent the server-rendered output of an LWC app:
+
+```html
+<body>
+    <x-content>
+        <template shadowroot="open">
+            <style>
+                main {
+                    background: #eee;
+                    padding: 1rem;
+                    border: 1px solid #aaa;
+                    border-radius: 0.25rem;
+                }
+            </style>
+            <h1>My article 2</h1>
+            <main>
+                <slot></slot>
+            </main>
+        </template>
+    
+        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
+        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
+        veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
+        commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
+        velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat
+        cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
+        est laborum.
+    </x-content>
+</body>
+```
+
+it renders correctly in Google Chrome but not in browsers that don't yet support Declarative ShadowDOM (e.g. Safari).
+
+## Motivation
+
+The primary goal of SSR is to get rendered content in front of the user as fast as possible, before enabling interactivity and eliminating all JS from SSR content. The Declarative Shadow DOM API was created to solve that goal (among others). The major browsers currently support the shadow dom API, but only Chrome supports the Declarative Shadow DOM API.
+
+LWC-SSR (`@lwc/engine-server`) generates its HTML output in compliance with the Declarative Shadow DOM spec, and for teams to take full advantage of SSR, declarative shadow dom must be polyfilled in those late adopter browsers.
+
+## Polyfills
+
+This section describes different terms and solutions to polyfill the Declarative ShadowDOM spec. It will define invariants for any of the described polyfill implementations, and compare them using the common comparison points when polyfilling the spec.
+
+### Invariants
+
+1. We would only support native shadow DOM when rendering declarative shadow DOM, synthetic shadow DOM will be out of the picture.
+2. The `getInnerHTML` method from the spec will not be polyfilled.
+3. The new `ElementInternals.shadowRoot` will not be polyfilled.
+4. For completeness, each polyfill removes the `<template>` element (and the inline `<script>` node when needed), so that the resulting tree is identical to the declarative output.
```

Comment:
Removing the `<template>` element is an implementation detail. I would suggest being a little more general for the invariant here.

```suggestion
4. The declarative shadow DOM polyfill should produce the same final DOM tree as its native counterpart.
```